### PR TITLE
[KlaudSol-CMS-#XX] Added JWT

### PR DESCRIPTION
**Pull request by JohnServo24**
Will be merged once `klaudsol-commons` is updated (hence, the build error).

The most relevant file: `/api/session.js`

I removed the handlers and the `try/catch` blocks on the API routes (except for `/login`), and I created an error handler on the frontend.

### CODE REVIEW

File: components/klaudsolcms/AppContentBuilderTable.js [hunk 1]
```diff
@@ -41,7 +42,7 @@ const AppContentBuilderTable = ({typeSlug}) => {
       (async () => {
         try {
           const valuesRaw = await slsFetch(`/api/entity_types/${typeSlug}/attributes/${attribute?.name}`, {
-            method: "DELETE"
+            method: "DELETE",
```
Comment (ardeearam) [inline components/klaudsolcms/AppContentBuilderTable.js:45]: What's the comma for?
```suggestion
            method: "DELETE"
```

File: jsconfig.json [hunk 2]
```diff
@@ -2,9 +2,9 @@
   "compilerOptions": {
     "baseUrl": ".",
     "paths": {
-      "@/*": ["/*"],
-      "@sme/*": ["/lib/sme/*"],
-      "@backend/*": ["/backend/*"]
+      "@/*": ["./*"],
+      "@sme/*": ["./lib/sme/*"],
+      "@backend/*": ["./backend/*"]
```
Comment (ardeearam) [inline jsconfig.json:7]: FYI, the direction is for us to make things simple and just remove all aliases except for the `"@/*": ["./*"]` one.

File: pages/admin/me.js [hunk 3]
```diff
@@ -7,9 +7,13 @@ import AppButtonSpinner from '@/components/klaudsolcms/AppButtonSpinner';
 import { FaCheck } from "react-icons/fa";
 import { useRef, useState } from 'react';
 import { slsFetch } from "@klaudsol/commons/lib/Client";
+import { useClientErrorHandler } from "@/components/hooks";
 import AppModal from '@/components/klaudsolcms/AppModal';
+import { useRouter } from "next/router";
 
 export default function Settings({cache}) {
+  const router = useRouter();
+  const errorHandler = useClientErrorHandler();
```
Comment (ardeearam) [inline pages/admin/me.js:16]: ❤️

File: pages/api/[entity_type_slug]/[id].js [hunk 4]
```diff
@@ -26,102 +26,73 @@ SOFTWARE.
...
-      case "DELETE":
-        return await del(req, res);
-      default:
-        throw new Error(`Unsupported method: ${req.method}`);
-    }
-  } catch (error) {
-    await defaultErrorHandler(error, req, res);
-  }
-}
+export default withSession(handleRequests({ get, del, put }));
```
Comment (ardeearam) [inline pages/api/[entity_type_slug]/[id].js:39]: Tight.

File: package.json [hunk 5]
```diff
@@ -33,6 +33,7 @@
     "@syncfusion/ej2-react-calendars": "^20.1.58",
     "aws-sdk": "^2.1098.0",
     "bootstrap": "^5.2.1",
+    "cookies-next": "^2.1.1",
```
Comment (ardeearam) [inline package.json:36]: Kindly bump up @klaudsol/commons version to:

"@klaudsol/commons": "https://github.com/klaudsol/klaudsol-commons.git#v1.1.0"

File: components/klaudsolcms/AppContentBuilderTable.js (hunk 1 above)
Comment (JohnServo24) [reply components/klaudsolcms/AppContentBuilderTable.js]: Oh, it's either my formatter, or when I removed the authorization headers, I accidentally removed the method and when I put it back, I entered a comma out of habit. Anyway, I removed the comma :)

File: components/errors/InvalidTokenError.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
+export default class InvalidTokenError extends Error {
```
Comment (ardeearam) [inline components/errors/InvalidTokenError.js:1]: Duplicate class: https://github.com/klaudsol/klaudsol-commons/blob/main/src/errors/InvalidTokenError.js

File: components/errors/MissingHeaderError.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
+export default class MissingHeaderError extends Error {
```
Comment (ardeearam) [inline components/errors/MissingHeaderError.js:1]: Duplicate class: https://github.com/klaudsol/klaudsol-commons/blob/main/src/errors/MissingHeaderError.js

File: components/klaudsolcms/modals/modal_body/CollectionTypeBody.js [hunk 8]
```diff
@@ -14,14 +18,15 @@ export default function CollectionTypeBody({ formRef }) {
       (async () => {
         try {
           //refactor to reducers/actions
-          await fetch(`/api/entity_types`, {
+          await slsFetch(`/api/entity_types`, {
```
Comment (ardeearam) [inline components/klaudsolcms/modals/modal_body/CollectionTypeBody.js:21]: Thanks :) I missed this part.

File: components/hooks.js [hunk 9]
```diff
@@ -20,4 +20,16 @@ export const useLogout = () => {
 export const useCapabilities = () => {
   const cache = useContext(CacheContext);
   return cache?.capabilities ?? [];
-};
\ No newline at end of file
+};
+
+export const useClientErrorHandler = () => {
```
Comment (ardeearam) [inline components/hooks.js:25]: Good call.

File: components/klaudsolcms/modals/modal_body/EditCollectionTypeBody.js [hunk 10]
```diff
@@ -29,15 +32,17 @@ export default function EditCollectionTypeBody({ formRef }) {
           //refactor to reducers/actions
           await fetch(`/api/entity_types/${slug}`, {
```
Comment (ardeearam) [inline components/klaudsolcms/modals/modal_body/EditCollectionTypeBody.js:33]: Can we go all the way with migrating `fetch` to `slsFetch`?

File: pages/api/entity_types/[slug]/attributes.js [hunk 11]
```diff
@@ -29,51 +29,34 @@ import { withSession } from '@klaudsol/commons/lib/Session';
 import { defaultErrorHandler } from '@klaudsol/commons/lib/ErrorHandler';
 import { OK, NOT_FOUND } from '@klaudsol/commons/lib/HttpStatuses';
 import { createHash } from '@/lib/Hash';
-import { setCORSHeaders } from '@klaudsol/commons/lib/API';
+import { setCORSHeaders, handleRequests } from '@klaudsol/commons/lib/API';
 import { assert, assertUserCan } from '@klaudsol/commons/lib/Permissions';
 import { readContentTypes, writeContentTypes } from '@/lib/Constants';
 
-export default withSession(handler);
+export default withSession(handleRequests({ post }));
```
Comment (ardeearam) [inline pages/api/entity_types/[slug]/attributes.js:36]: So clean, so good.

File: pages/api/session.js [hunk 12]
```diff
@@ -1,22 +1,30 @@
...
 import { assertUserIsLoggedIn } from '@klaudsol/commons/lib/Permissions';
 import EntityType from '@/backend/models/core/EntityType';
+import { setCookie } from 'cookies-next';
+import { generateToken } from '@klaudsol/commons/lib/JWT';
 
 export default withSession(handler);
 
 
 async function handler(req, res) {
+  // i hate cors
```
Comment (ardeearam) [inline pages/api/session.js:17]: 🤣

File: pages/api/session.js [hunk 13]
```diff
@@ -1,22 +1,30 @@
...
+  setCORSHeaders({ response: res, url: process.env.FRONTEND_URL });
```
Comment (ardeearam) [inline pages/api/session.js:18]: Let's prefix this with KS_
```suggestion
  setCORSHeaders({ response: res, url: process.env.KS_FRONTEND_URL ?? process.env.FRONTEND_URL });
```

**Pull request merged**